### DISCUSSION TITLE
Examples of a good fine-tune?

### DISCUSSION THREAD

**Discussion by addytheyoung** [yl4579/StyleTTS2#65] [General]
Does anyone have an example of a good fine-tuned styletts2 model? 

The only one I can find is the LJSpeech model, which sounds really good! But wondering what some other narrators / speakers would sound like, especially voices more outside the training dataset. Thanks, and awesome work on this.

**Comment by Shiro836** [comment]
https://www.youtube.com/watch?v=Tuz7_7q0Pr0 Trained on this interview: https://www.youtube.com/watch?v=ozOoONmJ9EQ

**Reply by addytheyoung** [reply]
Oh awesome! Was this trained on the LibriTTS checkpoint? Could you share the parameters you got to work? Ours is fine tuned on 1h of extremely clean narrator audio, cut into the same 10 second segments and adopts style but just gets really bad after 20 epochs, any idea what might be causing it?

https://voca.ro/1120thek63QO 33 epochs
https://voca.ro/11V6Bq9NVDEU 52 epochs

**Reply by Shiro836** [reply]
I used default Configs/config_ft.yml but with batch_size=2 and max_len=220, it's the maximum that could fit on my rtx 4090. I used Colab/StyleTTS2_Finetune_Demo.ipynb as finetuning script.
The important thing to be aware of is reference audio. After I used different reference audio for the same speaker, in my case, when I used audio from one of the forsen streams where he was loud and with a lot of background noise, the result was significantly worse.

**Reply by Shiro836** [reply]
I didn't change any other parameters, and training lasted for about an hour.

**Reply by Leemax0904** [reply]
Hi. Can you share how to generate the train_list.txt from the audio?

**Reply by Shiro836** [reply]
@CHIHHSIANGLI 
I used whisperx to transcribe and diarize. [My script](https://gist.github.com/Shiro836/a15eb97bb3f740a1324084e6aab36ada). The alternatives are [1](https://github.com/rioharper/VocalForge)(I didn't use it, but looks promising) [2](https://git.ecker.tech/mrq/ai-voice-cloning/)(When you finetune, this repo automatically creates dataset from your audio files)
After you have acquired audio to text dataset, you need to use phenomizer from the inference demo. [My script](https://gist.github.com/Shiro836/1c1881435ee75dc81069c505a14a9423) (it expects csv in format - first row: audio file path, second row: transcription)

**Reply by Shiro836** [reply]
I didn't use SLM adversarial training because I only have 24GB of vram. It's not enough for it.

**Reply by Oleksandr2505** [reply]
hi, have you converted your interview dataset on 24kzh before training? If so, How did you do this? what tools did you use? Thank you

**Reply by Shiro836** [reply]
@Oleksandr2505 shell script, input folder-wavs, output folder-tmp
 `for FILE in wavs/*; do ffmpeg -i $FILE -acodec pcm_s16le -ac 1 -ar 24000 tmp/$FILE -y; done`
you need ffmpeg for it

**Reply by slooi** [reply]
@Shiro836 Just wondering, how'd you capture Forsen's voice without the other dude's voice? Did you manually cut Forsen's voice out using ffmpeg or did you use some other tooling/automated software?

**Reply by Shiro836** [reply]
@slooi There is speaker diarization code at the end of the script I provided: https://gist.github.com/Shiro836/a15eb97bb3f740a1324084e6aab36ada. It uses https://huggingface.co/pyannote/embedding to get voice embeddings for speakers and differentiate between them.

**Reply by borrero-c** [reply]
@Shiro836 You might already be aware of this, but you can train the model with style diffusion in one fine-tuning session and adversarial training in another session. Should fit on your card since it ran fine on my 7900xtx. More details here under the SLM Adversarial Training section: https://github.com/yl4579/StyleTTS2/discussions/81.

I frankensteined your code with the code from [StyleTTS2FineTune](https://github.com/IIEleven11/StyleTTS2FineTune) to make my training data, using your code for the transcriptions and the other project for segmenting and phonemizing the data. Thank you for posting your scripts, it was extremely helpful having your code as a reference.

**Reply by GUUser91** [reply]
@Shiro836
I get this error when I used your script.
>ImportError: 'speechbrain' must be installed to use 'speechbrain/spkrec-ecapa-voxceleb' embeddings. Visit https://speechbrain.github.io/ for installation instructions.

**Reply by GUUser91** [reply]
If anyone has a graphics card with 24gb vram like the 7900 xtx or 4090, then it's possible to set the max_len settings to 280 if anyone has Linux. Just use virtual console mode and close any program eating up vram with nvtop.

**Reply by GUUser91** [reply]
@borrero-c
How were you able to do adversarial training? I followed the instructions by clicking on your link. I set the joint_epoch epoch to 48 so it can start on epoch 49 and adversarial training never started. I also followed the instructions on this link and adversarial training never started.
https://github.com/yl4579/StyleTTS2/issues/227#issuecomment-2069810056

**Reply by borrero-c** [reply]
@GUUser91 how many epochs are you training for? Are you doing two separate sessions or one?

You can train adversarial by using the diffusion trained model as the base model in the config.yml file, how are you doing it?

**Reply by GUUser91** [reply]
@borrero-c
I set the epoch to 80 for training. This is a seperate session. I'm using the diffusion checkpoint as the base model. I set joint_epoch each to 48 and then 49. Here is a screenshot of the training.
![IMG_20240427_113148_336](https://github.com/yl4579/StyleTTS2/assets/77461922/9863092e-277a-47e4-b17d-0f6020ebb7b2)

Edit:
I tinkered around with the config_ft.yml file. I set Max_Len to 120. I set batch_percentage to 1. I set slmadv_params min_len to 100 and slmadv_params max_len to 120. Batch size is set to 2. Now the DiscLM and GenLM Loss stats are no longer at 0. I'm using a rx 7900 xtx.
![Loss Training Screenshot_20240515_233529](https://github.com/yl4579/StyleTTS2/assets/77461922/84592cf4-9388-42c8-a073-e2b78aaab581)

Here a pic from my tensorboard folder.
![tensorboard update screencapture-localhost-6006-2024-05-15-23_49_50](https://github.com/yl4579/StyleTTS2/assets/77461922/b7d19cb8-a2f6-4391-9863-d2e6ebd03a46)

Another Edit: I discovered that SpectraLayers has a feature that can register and separate voices.

**Comment by yl4579** [comment]
This is the results I got using default config for 50 epochs (past `joint_epoch`) and one hour of data: https://vocaroo.com/1aC4vr4jErDL
If we do not run SLM adversarial training part (stop before `joint_epoch`) it is slightly worse: https://vocaroo.com/1hxkfwlrhowS

SLM adversarial training is the most VRAM consuming part. I don't know how to mitigate this problem and fit it in smaller machines. Maybe techniques used for LLM finetuning could help as we are working with large speech language models here.

**Comment by 78Alpha** [comment]
Here are the results I have from 2 different models. 

Aurora: 50 Epochs with joint training after 10, 8 Hours of audio, single voice. Batch Size 2, max length 220.
Chaos: 50 Epochs with joint training after 10, 10 hours of audio, 5 Hours of British audio for accent, 5 hours of depressed voice for the emotion. Batch Size 2, max length 220.

[AuroraTest1.webm](https://github.com/yl4579/StyleTTS2/assets/37676834/b365cdd4-7544-4076-a280-077f9fc970c5)
[AuroraTest2.webm](https://github.com/yl4579/StyleTTS2/assets/37676834/01246c34-407d-4cdd-a949-f237808c44a6)
[ChaosTest1.webm](https://github.com/yl4579/StyleTTS2/assets/37676834/0fa47927-8691-43a9-9fcd-a502eb164a5a)
[ChaosTest2.webm](https://github.com/yl4579/StyleTTS2/assets/37676834/49fa5303-45cb-4d91-8932-0dd5736a7456)

**Reply by godspirit00** [reply]
Hello, could you please share your config? Did SLM train during your finetuning? I was having a problem that even with `joint_epoch` small than `epochs`, SLM training did not start during the entire finetuning process. I wonder what was wrong. So it'd be great if you can share something about your finetuning experiment. Thanks.

**Reply by 78Alpha** [reply]
No joint training was done because I left batch percentage at 0.5. It needs to be at least 1 if you're using batch size 2.

**Reply by godspirit00** [reply]
I tried with batch percentage 1 and batch size 2, but it didn't start either.

**Comment by jonathandasilvasantos** [comment]
Fine-tuning on LibriTTS using a single Brazilian Portuguese speaker involved processing approximately 24 hours of audio over 60 epochs.

Link: https://drive.google.com/file/d/1pBqHbIuuaO7jvMsnnpbjrsFAPcHZKr41/view?usp=sharing

I'm using PL-BERT multilingual.

Please, any idea why there is this annoying noise on the end of the audio clip?

Thanks!

Jonathan S. Santos

**Comment by traderpedroso** [comment]
> Fine-tuning on LibriTTS using a single Brazilian Portuguese speaker involved processing approximately 24 hours of audio over 60 epochs.
> 
> 
> 
> Link: https://drive.google.com/file/d/1pBqHbIuuaO7jvMsnnpbjrsFAPcHZKr41/view?usp=sharing
> 
> 
> 
> I'm using PL-BERT multilingual.
> 
> 
> 
> Please, any idea why there is this annoying noise on the end of the audio clip?
> 
> 
> 
> Thanks!
> 
> 
> 
> Jonathan S. Santos 

Acredito que por falta de um pad de silêncio de pelo menos 400ms outra coisa se os áudios estiverem maior que o length faça o cálculo dos segundos e a frequência não tentei treinar ainda em português assim que concluir os LLM vou liberar um checkpoint em português se puder compartilhar seu check point

**Comment by GUUser91** [comment]
I tinkered around with the config_ft.yml file and I discovered I can do style diffusion and SLM Adversarial Training in one session on my 4090. Batch_size is set to 2. batch_percentage is set to 1.  Note this can also work on a 7900 xtx. I'm using Virtual console mode and I used nvtop to close any program eating up vram. Epoch is set to 100 because DiscLM is usually at 0. I'm using vokan as the base model. Most of the audio files I gathered had background music and noise so I used [resemble-enhance](https://github.com/resemble-ai/resemble-enhance) (Denoising via Gradio App version, not commandline version) and the audacity plug in acon digital deverberate 3 on the audio files. Then I used the audacity plug in [trim extend](https://github.com/nyancat18/audacity-plugin/blob/master/TrimExtend.ny) to add 200 milliseconds in the beginning and end of the audio files. 
Violet Parr: 8 minutes of audio. I set Max_Len to 270.  slmadv_params min_len is set to 270 and slmadv_params max_len is set to 270. 
https://vocaroo.com/15gtNvhjeFi0
Hiccup: 7 minutes of audio.  I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1b5ewbnxQ2aa
Branch: 11 minutes of audio. I set Max_Len to 256.  slmadv_params min_len is set to 256 and slmadv_params max_len is set to 256.  Model coudn't pronounce bouhuhuh-ned and buhuhuhuh.
https://vocaroo.com/1kwWiqOQuEvg
https://vocaroo.com/1f2epsXNHBtH
Poppy: 15 minutes of audio. I set Max_Len to 270.  slmadv_params min_len is set to 270 and slmadv_params max_len is set to 270. 
https://vocaroo.com/157F3j65bYLE
Arnold Shortman: 4 minutes of audio. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1ovdUEp7rVl2
https://vocaroo.com/1j7VQQSLo4Zo
Mr. Delicious: 4 minutes of audio. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1gJ3lSN6ut9t
Helga G. Pataki: 14 minutes of audio. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1bGX3i15t9Wm
https://vocaroo.com/1erNAm3ZH9n9
https://vocaroo.com/1cjvt2EHB6xm
Merida: 11 minutes of audio.  I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/118XYZ0Oq4t7
Judy Hopps: 21 minutes of audio.  I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1jf0jlUL3kOw
Wilbur Robinson: 7 minutes of audio. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. 
https://vocaroo.com/1agReLZAIx0c
Zuko: 6 minutes of audio. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280.  Model coudn't pronounce Katara.
https://vocaroo.com/1i9Sn1TvaPTr
https://vocaroo.com/1bkYnTkqhsyj
Connor: 56 seconds of audio taken from the film, Ruby Gillman, Teenage Kraken and 5 minutes of audio taken from Jaboukie Young-White (Connor's voice actor) interviews. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280.
https://vocaroo.com/17W0YDHnkgky
Mara Jade (Heidi Shannon): 2 minutes and 30 seconds of audio taken from the video game, Star Wars: Jedi Knight - Mysteries of the Sith and 5 minutes of audio taken from elevenlabs. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280.
https://vocaroo.com/1n04GhNXlon9
Luke Skywalker (Mark Hamill):  2 minutes of audio taken from a digital copy of Return of the Jedi and 7 minutes and 50 seconds taken from a 1983 interview. I set Max_Len to 280.  slmadv_params min_len is set to 280 and slmadv_params max_len is set to 280. Model coudn't pronounce Coruscant.
https://vocaroo.com/13KuwZNIMhuv

Edit: 
I rented out a h100 from runpod again. I edit the config_ft.yml file with micro. I set batch_size to 4 and max_len to 500. I set slmadv_params min_len to 100 and slmadv_params max_len to 500 and batch_percentage to 1. And now SLM adversarial training has started to work.
![Screenshot_20240604_194517](https://github.com/yl4579/StyleTTS2/assets/77461922/b4b16975-4685-4d6e-9661-d711a3e2ab0a)

Here's a screenshot of the vram usage.
![Screenshot_20240604_194444](https://github.com/yl4579/StyleTTS2/assets/77461922/f839e433-2058-44ed-8eb7-501b34d2c8ba)


This is what I did in runpod.
I update the repo.

>apt update

I install these.
>apt install aria2 p7zip-full curl jq micro

I use the pwd command to find directory / filepath infomation.
>pwd

I put the training dataset in a zip file and then I upload it to either https://catbox.moe/ or https://litterbox.catbox.moe/ (Which lets you upload a 1GB file)

I download the vokan base model and zip file with aria2.
>aria2c -x 16 -s 16 -k 1M https://archive.org/download/epoch_2nd_00012/epoch_2nd_00012.pth

or

>wget https://archive.org/download/epoch_2nd_00012/epoch_2nd_00012.pth

or the gofile downloader.

https://github.com/ltsdw/gofile-downloader.

Link for vokan model.
https://huggingface.co/ShoukanLabs/Vokan

>aria2c -x 16 -s 16 -k 1M https://files.catbox.moe/XXXXXXXX.zip

I unzip the file
>7z x XXXXXXXX.zip

I download the gofile upload script file.
>aria2c https://raw.githubusercontent.com/Sushrut1101/GoFile-Upload/master/upload.sh

I give the script permissions
>chmod +rx upload.sh

I upload the pth file to https://gofile.io/
>./upload.sh model.pth
